Billing: Show active payment method disabled for subscription when auto-renew is disabled

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This modifies the way that payment methods are displayed for a subscription when auto-renew is disabled. 

Originally, if a subscription had auto-renew disabled (barring other factors like an expired card), the payment method was always displayed. This was changed in https://github.com/Automattic/wp-calypso/pull/46625 because some customers were confused that the payment method was displayed even when it would not be used. However, it turns out this also causes confusion because now the page is lying (there is actually a payment method assigned, even though it's disabled); if auto-renew is turned back on, a payment method seems to appear out of nowhere. In addition, https://github.com/Automattic/wp-calypso/pull/46625 had the unintented side-effect of hiding the payment method if the card was expiring or had expired which also caused confusion.

In this PR we restore the original behavior of always displaying a payment method for a subscription with auto-renew disabled, but we grey-out the icon of that method to imply that the method will not be charged.

In addition, this adds unit tests to cover the various cases of how payment methods are displayed.

Fixes https://github.com/Automattic/wp-calypso/issues/48736

Before:

<img width="229" alt="purchase-info-before-narrow" src="https://user-images.githubusercontent.com/2036909/104952672-ed85d380-5992-11eb-811e-424f9a6c15ff.png">

After:

<img width="224" alt="purchase-info-after-narrow" src="https://user-images.githubusercontent.com/2036909/104952683-f080c400-5992-11eb-9b2f-cec915017679.png">


#### Testing instructions

- Visit the subscription page for a plan.
- Turn off auto-renew for the subscription.
- Click to add a payment method to that subscription.
- Return to the subscription page.
- Verify that auto-renew is still disabled, but that the payment method does not display "None".
- Verify that the payment method icon appears disabled.

sirbrillig committed:
Change renderPaymentInfo to component

sirbrillig committed:
Add isAutoRenewDisabled helper function

sirbrillig committed:
Prevent returning None for manualRenew purchase payment method

michaeldcain reviewed (approved):
This works as described - LGTM.

DavidRothstein commented:
I'm not sure if this will work as expected if auto-renew is disabled and the subscription expires soon?  I'll do some testing, but that's my initial reading of the code.

DavidRothstein reviewed (changes requested):
Yeah, I did some testing and if the subscription expires within 3 months, it still shows "None" with this pull request.  Probably need to use `isExpiring()` for all the checks, since that covers both of the cases where auto-renew is off (the data returned from the server is a bit confusing here).

I also noticed that `/me/purchases` is still showing subscriptions with auto-renew off as having no payment method.  We probably should keep that in sync with the data on this page also.

Overall I am very curious about how this behavior changed -- I really could have sworn that up until recently we always showed the payment method here.

sirbrillig commented:
> I'm not sure if this will work as expected if auto-renew is disabled and the subscription expires soon

Hm... that's a good point; this will cause us to skip the expiring check if auto-renew is disabled. Maybe we _should_ change the behavior of `isExpiring` to be what it says it is.

sirbrillig commented:
As we start to explore the edge cases here, I think this has become a prime opportunity for unit tests so we're clear what happens when. I'm going to add some.

DavidRothstein commented:
>Overall I am very curious about how this behavior changed -- I really could have sworn that up until recently we always showed the payment method here.

Looks like it was actually an intentional change from a few months ago: https://github.com/Automattic/wp-calypso/pull/46625. So I guess we should think a bit before undoing it.

Personally I think it's a bit confusing either way.  (Showing "None" kind of makes it look like your payment method is gone from the site when it isn't.)

DavidRothstein commented:
>Personally I think it's a bit confusing either way. (Showing "None" kind of makes it look like your payment method is gone from the site when it isn't.)

To elaborate on that a little more, we have a situation where your payment method says "None", but then you click on "Change Payment Method" (note, not "Add Payment Method" as would be expected if it didn't have a payment method) and the previously assigned payment method is pre-selected as the default.  Basically all the cues on that screen indicate that the payment method is still attached to your subscription (because it is), which contradicts with the "None" you saw elsewhere.

sirbrillig committed:
Fill in Purchase.payment type

sirbrillig committed:
Update return type of getByPurchaseId

sirbrillig commented:
> and the previously assigned payment method is pre-selected as the default

This is one reason why I disagree with the "pre-select the first payment method in the list" UX strategy. I think that even though the form is to "change the payment method", there's an implication that it's also displaying what's currently set.

sirbrillig commented:
> Looks like it was actually an intentional change from a few months ago: #46625

Thanks for finding that. I looked all over the blame and somehow missed that change 🤦

sirbrillig commented:
I agree that the UX without this change is awkward, but correct me if I'm wrong: having a payment method attached to a subscription when that subscription has auto-renew disabled does absolutely nothing, right? 

If that's correct, then another solution to the UX might be to disable the "Add a payment method"/"Change a payment method" button if autorenew is disabled, since it's a noop action (unless, as we have discussed in https://github.com/Automattic/wp-calypso/issues/48736, adding a payment method re-enables autorenew).

sirbrillig commented:
> Showing "None" kind of makes it look like your payment method is gone from the site when it isn't.

Sorry for all the piled up questions, but what if instead of "None", this read, "Autorenew is disabled"?

DavidRothstein commented:
Something like "autorenew is disabled" might help a bit.  I think there are still some reasons to show the currently-assigned payment method itself though.  In the vast majority of cases, the currently-assigned payment method is the one the user paid with last time, and it's useful to see that.  Also, we don't currently do this, but you could imagine selecting that payment method by default in checkout when they go to manually renew.  Finally, it just seems weird to me that toggling auto-renew on can magically make a payment method appear that wasn't there before.

To me it seems like we would ideally show the payment method, but deemphasized in some way (like faded out or something).  The existing "auto-renew is off" message should be the most prominent.   But I'm not a designer :)

If we **are** going to keep treating it as not having a payment method, though, maybe we should do that in the data returned from the API endpoint.  That way we'd actually treat it as not having a payment method everywhere in Calypso.  (That would solve the issue with it saying "Change payment method" rather than "Add payment method" and various other things.)

sirbrillig commented:
After some additional discussion echoing the above, we decided to display the payment method instead of "None" if autorenew is disabled, but make the payment method appear "disabled" (with a grey overlay perhaps? something similar to the display of an unselected payment method in checkout) in that case. Working on that...

We also need to update the account-level view.

sirbrillig committed:
Move "Payment method" header and block to PaymentInfoBlock

sirbrillig committed:
Add aria-label to PaymentInfoBlockWrapper

sirbrillig committed:
Add basic test for PaymentInfoBlock

DavidRothstein commented:
In addition to the disabled styling, I wonder if it would also be helpful to change the existing "Auto-renew is OFF" text to something more like "Auto-renew is OFF (your payment method will not be billed)" in this scenario, to emphasize it.

That might also better complement the text that is currently displayed when auto-renew is on, which currently includes "You will be billed on..." wording.

sirbrillig committed:
Add aria-label for placeholder in PaymentLogo

sirbrillig committed:
Add more tests for PaymentInfoBlock

sirbrillig committed:
Add tests for expiring and expired cards

sirbrillig committed:
Add tests for when auto-renew is disabled

sirbrillig committed:
Display expired payment method if auto-renew is disabled

Since both use `expiryStatus` as their field, this should never be
possible, but it makes the logic more clear.

sirbrillig commented:
> Yeah, I did some testing and if the subscription expires within 3 months, it still shows "None" with this pull request. Probably need to use isExpiring() for all the checks, since that covers both of the cases where auto-renew is off (the data returned from the server is a bit confusing here).

I... think this is intended behavior? That is, if you have an attached credit card which is expiring, it will display "None" if auto-renew is on or off. Am I misunderstanding?

DavidRothstein commented:
I don't think it's expected.  If your subscription is going to renew soon, that's one of the most likely times you'd be looking at this page.  Why should the payment method disappear then?

(Note that the scenario I'm talking about is when the _subscription_ expires soon, not when the credit card expires soon -- although the same argument applies for both.)

sirbrillig commented:
> (Note that the scenario I'm talking about is when the subscription expires soon, not when the credit card expires soon -- although the same argument applies for both.)

Ah! Sorry, I misunderstood. As for the credit card expiring soon, the existing code is quite explicit that it displays "None" in that case, and while I assume that might have been intentional I agree that it's not very clear. I'm happy to modify it here (and we have tests now which should make it easier!). Do you have a suggestion for what should happen in that case? Something like... display the payment method but with "Expires soon" text included?

DavidRothstein commented:
>As for the credit card expiring soon, the existing code is quite explicit that it displays "None" in that case

Where do you see that?

When auto-renew is on, it displays the credit card for me, even if it's expiring soon.  And in that case the message at the top of the page is what tells you about the potential issue with the card:

![credit-card-expiring-soon](https://user-images.githubusercontent.com/235183/104520478-33285200-55c9-11eb-930c-1d371c01b6d6.png)

I think if auto-renew is off we can display it also (same as for any other situation; if auto-renew is off, it shouldn't really matter if the credit card is expiring soon, since that only comes into play if they toggle auto-renew back on).

sirbrillig committed:
Add disabled prop to PaymentLogo

sirbrillig committed:
Disable PaymentLogo when auto-renew is disabled

sirbrillig committed:
Add test for disabled payment logo

sirbrillig commented:
> Where do you see that?

It's here:

https://github.com/Automattic/wp-calypso/blob/1a73d6aa5f36b407580a8c1ce48507457434ae68/client/me/purchases/manage-purchase/purchase-meta.jsx#L239-L241

That's weird that you can still see it... 🤔

sirbrillig committed:
Add test for paypal

sirbrillig committed:
Make sure PaymentLogo is always rendered if there is a payment method

sirbrillig committed:
Add tests to make sure payment methods are displayed when expired

sirbrillig committed:
Remove expired/expiring check in PaymentInfoBlock

sirbrillig commented:
Ok, I've updated the PR description to better explain what this PR does. It effectively reverts https://github.com/Automattic/wp-calypso/pull/46625 but makes the logo grayscale if auto-renew is disabled. This will be more obvious for big logos like PayPal but less obvious for the small credit card logos. 

I personally feel like we need to also add some text in there to be explicit (eg: for screen readers and less-color-sensitive users) but I don't know what it should say.

DavidRothstein commented on client/me/purchases/manage-purchase/payment-info-block.tsx:
```diff
@@ -0,0 +1,78 @@
+		if ( isPaidWithCreditCard( purchase ) ) {
+			return (
+				<PaymentInfoBlockWrapper>
+					<PaymentLogo type={ logoType } disabled={ isAutoRenewDisabled( purchase ) } />
```
I believe all these should be `isExpiring()` rather than `isAutoRenewDisabled()`.  Otherwise the styling will be wrong when the subscription is close to expiration.

There shouldn't actually be a need to introduce `isAutoRenewDisabled()` in this pull request, since it's covered as part of `isExpiring()`.

sirbrillig replied:
Ah, _now_ I understand what you were saying above about using `isExpiring`. I was very confused because I thought it was talking about the credit card.

DavidRothstein commented:
>It's here:
>
>https://github.com/Automattic/wp-calypso/blob/1a73d6aa5f36b407580a8c1ce48507457434ae68/client/me/purchases/manage-purchase/purchase-meta.jsx#L239-L241
>
>That's weird that you can still see it... 🤔 

That code is for expiring subscriptions (i.e. subscriptions that expire soon), not credit cards that expire soon.  So it all checks out :smile: 

>I personally feel like we need to also add some text in there to be explicit (eg: for screen readers and less-color-sensitive users) but I don't know what it should say.

Definitely; I tried it out and it is pretty subtle now.  What did you think of my suggestion [above](https://github.com/Automattic/wp-calypso/pull/48775#issuecomment-759604679) to add more text to the "Auto-renew is OFF" section?  Or maybe the same "your payment method will not be billed" text I suggested there could go underneath the payment method instead, to make it more visible -- not sure which is best.

sirbrillig committed:
Clarify in test cases that expiryStatus is for purchase not card

sirbrillig committed:
Disable PaymentLogo when isExpiring instead of isAutoRenewDisabled

sirbrillig committed:
Remove unused isAutoRenewDisabled

sirbrillig committed:
Add test for extra text displayed when autorenew is disabled

sirbrillig committed:
Add "will not be billed" text when auto-renew is disabled

sirbrillig committed:
Remove nowrap on purchase date

It was causing the text to overflow the box.

sirbrillig commented:
Updated with "(this payment method will not be charged)" text. Let me know what you think. I also updated the screenshots in the description.

sirbrillig commented:
Sorry it took me so long to grok `isExpired` 😅

DavidRothstein commented on client/me/purchases/manage-purchase/payment-info-block.tsx:
```diff
@@ -0,0 +1,101 @@
+		<aside aria-label={ String( translate( 'Payment method' ) ) }>
+			<em className="manage-purchase__detail-label">
+				<span>{ translate( 'Payment method' ) }</span>
+				{ isExpiring( purchase ) && (
```
I think you need some more conditions here -- for example, we don't want to display "this payment method will not be billed" when the payment method is "None" (and perhaps not when the payment method is "Credits" either?).

DavidRothstein commented on client/me/purchases/manage-purchase/payment-info-block.tsx:
```diff
@@ -0,0 +1,101 @@
+			<em className="manage-purchase__detail-label">
+				<span>{ translate( 'Payment method' ) }</span>
+				{ isExpiring( purchase ) && (
+					<div className="manage-purchase__detail-label-subtitle">
```
Technically this might be invalid HTML (`<div>` inside `<em>`)?

sirbrillig replied:
Yeah... you're right. I considered that. What I _wanted_ was for there to be no margin between the title (which should probably be an `h` tag of some level rather than an `em` since it's made a block-level element by the CSS) and the subtitle, but I want the subtitle to also be a block-level element. I had originally done that with CSS, but then I realized that I may as well just use HTML that made it a block-level element. However, this results in the issue you describe. I'll refactor the `em` and the subtitle to hopefully resolve this.

DavidRothstein commented on client/me/purchases/manage-purchase/style.scss:
```diff
@@ -313,7 +313,3 @@
-
-.manage-purchase__detail-date-span {
-	white-space: nowrap;
-}
```
I'm not sure why this is being removed -- to me it looked much better before.

Before:

![before](https://user-images.githubusercontent.com/235183/104878288-a31c3c80-5929-11eb-83c7-ea8df04317d9.png)

After:

![after](https://user-images.githubusercontent.com/235183/104878286-a31c3c80-5929-11eb-90ef-c4c185eaf2b3.png)

sirbrillig replied:
But have you tried it at smaller widths?

<img width="269" alt="Screen Shot 2021-01-18 at 11 54 21 AM" src="https://user-images.githubusercontent.com/2036909/104943568-ee633900-5983-11eb-8209-678fe0c82e10.png">

DavidRothstein replied:
Hm, I had tried it at the smallest (mobile) widths, where it looked OK, but hadn't tried intermediate widths.  However, it seems to really be a larger problem with the page design not working at those widths.  The date is not the only thing that spills over:

![medium-width](https://user-images.githubusercontent.com/235183/104958147-7dc91600-599d-11eb-9ca8-34060a0bf551.png)

Personally I would leave this code as is; I don't think the tradeoff of making it look worse on wide screens is worth it.  Or, could you change this styling to only apply at a large-screen-width breakpoint?

DavidRothstein reviewed (commented):
This looks good (and nice tests!).  And I think displaying the extra text inside the payment method section does work pretty well.

I think the text could simply be "will NOT be billed", though (rather than "this payment method will not be billed"), to avoid repeating words.  And putting NOT in all capital letters would match how OFF is in all capital letters in the Subscription Renewal section which I think might make it easier to understand, although I don't feel too strongly about that.

I left a few inline comments also.

sirbrillig commented:
> I think the text could simply be "will NOT be billed", though (rather than "this payment method will not be billed"), to avoid repeating words. And putting NOT in all capital letters would match how OFF is in all capital letters in the Subscription Renewal section which I think might make it easier to understand, although I don't feel too strongly about that.

I 100% agree if this were english-only. However, I wonder how all-caps words translate to other languages. "will NOT be billed" is using character case as an emphasis, which is a locale-dependant concept, probably better handled by style than by capitalization. Also, it has an implicit subject, which is probably fine (especially with appropriate translators notes) but my instinct is to be explicit. Maybe we could compromise with something like "this will not be billed"? That makes the text a complete sentence, but uses "this" for the subject rather than "this payment method", and the context of "this" should be apparent by the placement of the text.

sirbrillig committed:
Update text to check only for the second half of "will not be billed"

sirbrillig committed:
Change non-auto-renew text to "this will not be billed"

sirbrillig committed:
Move "will not be billed" text to own block

sirbrillig committed:
Update tests to check "will not be billed" only for some payment methods

sirbrillig committed:
Only show "will not be billed" for CC/PayPal

sirbrillig commented:
Updated to make sure "will not be billed" is only rendered for CC or PayPal.

DavidRothstein reviewed (commented):
I don't know if the "this" is necessary (translators should be able to figure out how to handle it) but it also looks fine to me.  You are right about capital letters not being great; I was mainly just suggesting that the style should match the other section (which already uses capital letters) but I think it's OK without that too.

I took a quick look at how this all works with Stripe Source payment methods such as iDEAL.

Initial screen:

![ideal-manage-purchases](https://user-images.githubusercontent.com/235183/104958536-33946480-599e-11eb-892a-d12620b6f209.png)

Trying to turn auto-renew on (note that these wind up as auto-renew off always, even right after purchase):

![ideal-toggle-auto-renew-on](https://user-images.githubusercontent.com/235183/104958543-35f6be80-599e-11eb-82a3-d1be823ade76.png)

Perhaps not ideal (ha ha, get it?) -- maybe would be better to continue treating those like "None", at least for now?

sirbrillig committed:
Add test for iDEAL payment method

sirbrillig committed:
Fix boolean prop type in PaymentInfoBlockWrapper

sirbrillig committed:
Add isRechargeable to Purchase type

sirbrillig committed:
Clarify PurchasePayment type that props can be undefined

sirbrillig committed:
Make sure "no payment method" test still has a purchase object

sirbrillig committed:
Show "None" for non-rechargable payment methods

sirbrillig committed:
Revert removal of white-space: nowrap

sirbrillig commented:
Updated to always display "None" for non-renewable payment methods like iDEAL and to revert the CSS change.

DavidRothstein reviewed (approved):
The latest changes look good, and the various payment methods I tested all seem to work as expected now.  Thanks!

sirbrillig committed:
Move PaymentInfoBlock to own file

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5466846

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5466846).
